src/Controller/Admin/Document/DocumentController.php

```diff
@@ -1088,45 +1099,57 @@ public function diffVersionsAction(Request $request, int $from, int $to): Respon
         }
 
-        $prefix .= $docFrom->getRealFullPath() . '?pimcore_version=';
+        $comparisonId = uniqid(date('Y-m-d') . '-', true);
+        $tempFileTemplate = PIMCORE_SYSTEM_TEMP_DIRECTORY . '/version-diff-tmp-' . $comparisonId . '-%s.%s';
+        $fromImageFile = sprintf($tempFileTemplate, 'from', 'png');
+        $toImageFile = sprintf($tempFileTemplate, 'to', 'png');
+        $diffImageFile = sprintf($tempFileTemplate, 'diff', 'png');
+        $fromHtmlFile = sprintf($tempFileTemplate, 'from', 'html');
+        $toHtmlFile = sprintf($tempFileTemplate, 'to', 'html');
 
-        $fromUrl = $prefix . $from;
-        $toUrl = $prefix . $to;
+        $viewParams = [];
 
-        $toFileId = uniqid();
-        $fromFileId = uniqid();
-        $diffFileId = uniqid();
-        $fromFile = PIMCORE_SYSTEM_TEMP_DIRECTORY . '/version-diff-tmp-' . $fromFileId . '.png';
-        $toFile = PIMCORE_SYSTEM_TEMP_DIRECTORY . '/version-diff-tmp-' . $toFileId . '.png';
-        $diffFile = PIMCORE_SYSTEM_TEMP_DIRECTORY . '/version-diff-tmp-' . $diffFileId . '.png';
+        $docContentFrom = $documentRenderer->render($docFrom);
+        $docContentTo = $documentRenderer->render($docTo);
 
-        $viewParams = [];
+        file_put_contents($fromHtmlFile, $docContentFrom);
+        file_put_contents($toHtmlFile, $docContentTo);
 
-        $session = $request->getSession();
+        $prefix = Config::getSystemConfiguration('documents')['preview_url_prefix'];
+        if (empty($prefix)) {
+            $prefix = $request->getSchemeAndHttpHost();
+        }
 
-        Chromium::convert($fromUrl, $fromFile, $session->getName(), $session->getId());
-        Chromium::convert($toUrl, $toFile, $session->getName(), $session->getId());
+        try {
+            Chromium::convert($prefix . $router->generate('pimcore_admin_document_document_diff_versions_html', ['id' => basename($fromHtmlFile)]), $fromImageFile);
+            Chromium::convert($prefix . $router->generate('pimcore_admin_document_document_diff_versions_html', ['id' => basename($toHtmlFile)]), $toImageFile);
+        } finally {
+            unlink($fromHtmlFile);
+            unlink($toHtmlFile);
+        }
 
-        $image1 = new Imagick($fromFile);
-        $image2 = new Imagick($toFile);
+        $image1 = new Imagick($fromImageFile);
+        $image2 = new Imagick($toImageFile);
 
         if ($image1->getImageWidth() == $image2->getImageWidth() && $image1->getImageHeight() == $image2->getImageHeight()) {
             $result = $image1->compareImages($image2, Imagick::METRIC_MEANSQUAREERROR);
             $result[0]->setImageFormat('png');
 
-            $result[0]->writeImage($diffFile);
+            $result[0]->writeImage($diffImageFile);
             $result[0]->clear();
             $result[0]->destroy();
 
-            $viewParams['image'] = $diffFileId;
+            $viewParams['image'] = base64_encode(file_get_contents($diffImageFile));
```

Comment:
Overall LGTM, just noticed this part, instead of `writeImage`, could be instead passing an `getImageBlob` to avoid a temp file?

```suggestion
            $viewParams['image'] = base64_encode($result[0]->getImageBlob());
```

Reply:
@kingjia90 Yep, makes sense - pushed the change -thx